### PR TITLE
fish_git_prompt: fix macOS lag after reboot and problem when Xcode not installed

### PR DESCRIPTION
**Pull request by floam**
## Description

This fixes the two xcrun-related problems encountered due to the nature of the `/usr/bin/git` shim on macOS:

### Slow `git` after reboot

Due to $TMPDIR being cleaned out on reboots, and a cache file that is slow to generate, it can take multiple seconds for the `git` shim to locate the actual Git binary inside of the developer directory that came with Xcode/CLT, and it does some security checks on the the Xcode.app bundle that can be slow before creating a cache file. We detect the file not existing, and run `git --version &` in the background, skip printing the git prompt, and when an on-job-exit handler for that `git --version` finally runs we reprint the prompt. 

The behavior is nice. You can compare what happens after `rm (xcrun --show-cache-path)`: instead of fish appearing to hang for a while before the prompt finally prints, you get the asynchronous prompt effect with a regular prompt that displays the git portion later.

Fixes issue #6625

### Git might not actually be installed

If Xcode or the smaller command line tools package aren't actually installed or working, running `git` would cause a dialog to appear asking the user to install it. This isn't nice if you are running that binary every time you draw the prompt.

Fixes issue #9343

### CODE REVIEW

**Comment by floam** [comment]
I notice the problem isn't as bad as it used to be, the slow cache regeneration. It takes like 900ms on my current M1 Pro  MacBook Pro. It definitely used to be 5, 10+ seconds on older hardware; It's possible Xcode has gotten a little faster over the years too (xcrun has been a thing since, oh 2012/2013 Mavericks I believe.)

**Review by ridiculousfish — CHANGES_REQUESTED** [review]
Man it suck so hard we have to do this. But we do, you're right. Thank you for tackling it.

I think your overall approach is right. I have some suggestions for simplifying it.

Instead of polluting `fish_git_prompt`, let's introduce a new function (maybe `__fish_git_prompt_ready`) whose job it is to figure out if it's safe to run git. That's where we'll add this logic on Mac.

Now instead of adding this top-level `uname` variable we can just define the function conditionally:

    if "$(uname)" = Darwin -a "$(command -v git)" = /usr/bin/git
             function __fish_git_prompt_ready
                        # all the stuff
             end
    else
            function __fish_git_prompt_ready
                  return 0
            end
    end
    
Second, I like the approach of running `git` in the background and ensuring we only do it once (if something goes wrong, we don't want to spawn a bunch of background procs). However I think we can achieve the same by simply running it at global scope. The extra waiting function can also be omitted; we'll just test for the path.

So basically the structure I propose is:

    if "$(uname)" = Darwin -a "$(command -v git)" = /usr/bin/git
             # Kick off git in the background to populate the cache.
             command git --version &>/dev/null 2>
             function __fish_git_prompt_ready
                        test -e "$(xcrun --show-cache-path)" || return 1
                        # Path exists, replace our function.
                        function __fish_git_prompt_ready; return 0; end
                        return 0
             end
    else
            function __fish_git_prompt_ready
                  return 0
            end
    end

**Comment by ridiculousfish** [comment]
[An article about this nonsense](https://lapcatsoftware.com/articles/xcrun.html)

**Comment by floam** [comment]
The reason the security checks are done, is somethuing we might vaguely remember, [XcodeGhost](https://en.wikipedia.org/wiki/XcodeGhost).

**Comment by floam** [comment]
After sleeping on it,I realize there is going to be an additional cold-start "install on demand" issue: python3, when we do that one-off thing with the manpage completion generator script. (macOS no longer comes with Python 2, Python 3 is shipped with the command line tools.)

**Comment by floam** [comment]
To reproduce slowness issue:
```fish
rm (xcrun --show-cache-path)
```

To reproduce the virgin macOS issue:
```fish
sudo rm -rf /Library/Developer/CommandLineTools
sudo mv /Applications/Xcode.app/ /Applications/Xcode_.app
sudo xcode-select --reset
```

**Comment by ridiculousfish** [comment]
@floam do you anticipate getting to this shortly, or shall I take it over?

**Comment by floam** [comment]
I was a bit busy but things should clear up tomorrow. I’ll get into it.

**Comment by faho** [comment]
>For the manpage completions, I'd prefer honestly just skipping what we do on macOS and, if python3 exists, installing the equivalent of a monthly cron job into launchd on macOS. I don't dig how it never re-runs.

We don't do that anywhere else, and adding a scheduled job that runs outside of fish (whether cron or something else) is a bit of a land grab that I'm uncomfortable with. For one we would also have to ensure it's removed when fish is.

It's also not typically much of an issue - the list of commands doesn't change by a large percentage (the long tail of simple commands is already gonna be part of the initial install), and man completions aren't great anyway.

So I would be against this. A compromise I would be okay with is documenting these things and letting the user enable them (e.g. shipping a disabled systemd timer and the launchd equivalent would be okay by me).

**Comment by ridiculousfish** [comment]
I'm proposing https://github.com/fish-shell/fish-shell/pull/9434 which builds on these ideas. It does not include any sort of cron job or launchd agent, it merely skips running python3 by default if it's only a stub.
